Default: Prevent placing sapling if grown tree intersects protection

paramat opened the pull request:
Add a global 'intersects protection' function to functions.lua for
checking if a specified volume intersects with a protected volume.
A 3D lattice of points are checked with an adjustable interval.
Add a global 'sapling on place' function to avoid duplicated code in
nodes.lua.
/////////////////////////////////////////////////////

Action for issue #1192 

A function is added to check protection throughout a specified volume. Check points are a 3D lattice with a specified interval (here being 4), the lattice is automatically shrunk independently along each axis so that the corners and borders contain check points. Each check point then calls 'is_protected()'.

For each sapling, placement is refused if either the node containing the saplng is protected or the resulting volume of the grown tree intersects protection (in which case an extra message explains why placement failed). The tree volume and check interval is easiily set in a sapling's node def 'on place' function.

Mod trees can use these new global functions, and the check function is useful for many uses such as schematic placement.

Interval 4 could occasionally miss 3x3x3 node protected areas, but these are very rare. Most trees will have 12 checks, jungletree 16, acacia 27. Checks throughout the volume are needed because trees are tall (and mod trees can be huge) so just checking volume corners can easily miss a protected volume within those corners.

Initial discussion http://irc.minetest.ru/minetest-dev/2016-07-12#i_4648623
Development and simplification http://irc.minetest.ru/minetest-dev/2016-07-14#i_4650224
Further development and discussion below.

A serious griefing issue, from IRC project channel http://minetest.org/ircproject/2016-07-14.html

04:15:10 <Shara> Red Cat has also had severe tree griefing issues
04:17:09 <DonBatman> Shara and VanessaE Was the tree griefing from below?
04:17:18 <Shara> Yes
04:17:28 <Shara> People digging under buildings to do it
04:17:36 <VanessaE> DonBatman: I don't know what the case was on my server, one of my admins reported it and already started cleaning up the damage
04:18:41 <Shara> Someone logged in around about the time I left for work, and seemed to spend a full day trying to wreck as many buildings as possible with it
04:20:04 <Shara> They used the tallest trees possible and dug under protection.
04:20:24 <Shara> Whole place was basically a solid clump of wood and leaf
04:20:41 <Shara> Players signing in for the first time couldn't even leave spawn

Rogier-5 commented:
@paramat Several moretrees trees have diameters up to 20 nodes, some can even be a bit larger. So it makes sense checking at horizontal intervals too. As there is a tradeoff here, I think it would be a good idea to to make the interval configurable. Few people will need to change the default, but those who do will be grateful.

IMO a default of 5 nodes seems OK, few if any builds will probably be that small. But that will still result in 250-300 checks for a sequoia sapling. Performance-consious server operators might prefer an interval of 10 nodes, reducing the number of checks for a sequoia sapling to 50-60, and accepting that smaller builds can still be griefed.

As most protected areas will probably be larger than just a few nodes, an optimisation could be to check the corners of the area first, then only then check the interior. In most cases, this will reduce the number of checks to at most 8. This has the added benefit that all of the corners will actually be checked. Currently, the 4 bottom-most corners can remain unchecked.

The function `intersects_protection()` might be made part of the protection api, allowing a mod to replace it with its own, more efficient implementation.

Rogier-5 commented on mods/default/nodes.lua:
```diff
@@ -500,16 +500,36 @@ minetest.register_node("default:sapling", {
+		local player_name = placer:get_player_name()
+		if not default.intersects_protection(
+				-- minp, maxp of resulting tree at maximum height
+				{x = pos.x - 2, y = pos.y - 1, z = pos.z - 2},
```
The sapling is placed at `pos`, and the tree will grow from `pos` up. `pos.y-1` will not be affected, and any other node *could* be placed at `pos`. Why prevent placing the sapling ?

paramat replied:
I agree, mapgen trees have a root but sapling-grown ones don't.

paramat commented:
All good points, will work on it.

Ekdohibs commented:
> The function intersects_protection() might be made part of the protection api, allowing a mod to replace > it with its own, more efficient implementation.
This won't work, because multiple mods may add protection. It is not possible to keep efficiency with the current API, which thus has to be redesigned. Anyway, this is a discussion for the engine instead.

paramat commented:
Updated with suggestion.
Check corners first, then an interior 3D lattice at specified intervals.

Rogier-5 commented on mods/default/functions.lua:
```diff
@@ -481,3 +481,35 @@ minetest.register_abm({
+	for z = minp.z + interval, maxp.z - 1, interval do
+	for y = minp.y + interval, maxp.y - 1, interval do
+	for x = minp.x + interval, maxp.x - 1, interval do
+		if minetest.is_protected({x = x, y = y, z = z}, name) then
```
I think these loops do not visit the borders of the area, other than at the corners. E.g. `{ x = minp.x + 5, y = minp.y, z = minp.z + 5}` will not be vistited. This would
allow placing a sapling below a protected area near to its border, for a tree which will extend to above the protected area. As the borders of the tree-area are only checked exactly above and below the protected area, and not in the middle of the area, this would allow leaves in the middle to grief the protected area.

Example: a sapling for the following tree would be placeable:
```
LLLLLLLLLLL
LLLLLL|---------------------
LLLLLL|LLLL
LLLLLL|LLLL
LLLLLL|LLLL
LLLLLL|LLLL
LLLLLL|LLLL   protected
     T|        area
     T|
     T|
     T|-----------------------
     T
```

paramat replied:
Yeah well spotted.

Rogier-5 replied:
It looks like a slightly modified form of my example above is still possible. Assuming x increases from left to right:
```
LLLLLL|LL        (protected area on this side)
^--------------- minp.x                                 checked: not protected
     ^---------- minp.x + 1 * interval                  checked: not protected
          ^----- minp.x + 2 * interval > maxp.x         not checked
      ^^^------- griefed
```
Using interval checking, it will never be possible to catch everything, but I think this is a case that ought to be caught. Else every protected area would have to be made higher than any tree could be (which apparently is about 55 nodes for moretrees atm)

paramat commented:
Updated.

Rogier-5 commented on mods/default/functions.lua:
```diff
@@ -481,3 +481,35 @@ minetest.register_abm({
+
+	for z = minp.z, maxp.z, interval do
+	for y = minp.y, maxp.y, interval do
+	for x = minp.x, maxp.x, interval do
```
How about code like the following (disclaimer: not tested) for avoiding re-checking the corners. Assuming a protection check can be expensive, and assuming the volume to check is not too large, this would be more efficient:
```
-- Avoid re-checking the corners
-- (assuming protection checks can be expensive,
--  and the checked volume is not too large.)
local corner = 0
corner = corner + ((x == minp.x or x == maxp.x) and 1 or 0)
corner = corner + ((y == minp.y or y == maxp.y) and 1 or 0)
corner = corner + ((z == minp.z or z == maxp.z) and 1 or 0)
if corner < 3 then
	-- do protection check
end
```

Rogier-5 replied:
> this would be more efficient

Assuming of course that the loops check all boundaries of the volume, which currently is not necessarily the case.

paramat replied:
The corners are usually not rechecked because interval is rarely a perfect division of volume dimensions.

paramat replied:
However will think on this.

Rogier-5 commented on mods/default/functions.lua:
```diff
@@ -481,3 +481,35 @@ minetest.register_abm({
+			minetest.is_protected({x = maxp.x, y = minp.y, z = minp.z}, name) or
+			minetest.is_protected({x = maxp.x, y = minp.y, z = maxp.z}, name) or
+			minetest.is_protected({x = maxp.x, y = maxp.y, z = minp.z}, name) or
+			minetest.is_protected({x = maxp.x, y = maxp.y, z = maxp.z}, name) then
```
Why not check the eighth corner ?

Rogier-5 replied:
Oh sorry. Forgot to read your comment.

paramat commented:
Tested by Shara as ok.

paramat commented:
Updated to not double-check corners.
Code now applied to all saplings.

t4im commented on mods/default/nodes.lua:
```diff
@@ -500,16 +500,39 @@ minetest.register_node("default:sapling", {
+					-- interval of interior volume check 
+					4
+				) then
+			minetest.set_node(pos, {name = "default:sapling"})
```
This will likely replace any unprotected item, if it is `pointed_thing.above` and ignores any `buildable_to` `pointed_thing.under`

paramat replied:
I see it's not necessarily checking the sapling pos.
Good points.

Rogier-5 commented:
I tested this using moretrees (after patching it to support protection checking).

I tried placing saplings to the east and to the west of a 100x30x100 protected area, both at the same level as the protected area, and one node below.

Placing a sapling to the east is denied when too close to the protected area, whether the sapling is placed at the same level as the protected area, or below.

Placing a sapling to the west is denied at the correct distance when the sapling at the same level as the protected area, but when placing it one level lower, the sapling could be placed 3 nodes closer to the protected area, which resulted in griefing 3 nodes deep into the protected area. This is sufficient to cause leaves to appear on the other side of a wall, inside a protected building.

Although I was able to 'fix' this by tweaking the interval and the protection distance, IMHO such tweaking is hard to verify, easy to forget or get wrong, and not future-proof (as it creates a dependency between of sapling code on the implementation of the protection checking code).

I made some changes to get it to work as expected, and I'd like to propose the following code which fixes it, and prevents placing saplings too close to a protected area, regardless whether it is to the east, west, north, south, or whether the sapling is placed at a lower level than the protected area or not. It does this by adjusting the interval, so that all borders of the area are always checked.

```
function default.intersects_protection(minp, maxp, name, interval)
        local corner1 = {x = minp.x, y = minp.y, z = maxp.z}
        local corner2 = {x = minp.x, y = maxp.y, z = minp.z}
        local corner3 = {x = minp.x, y = maxp.y, z = maxp.z}
        local corner4 = {x = maxp.x, y = minp.y, z = minp.z}
        local corner5 = {x = maxp.x, y = minp.y, z = maxp.z}
        local corner6 = {x = maxp.x, y = maxp.y, z = minp.z}
        local corner7 = {x = maxp.x, y = maxp.y, z = maxp.z}

        if minetest.is_protected(corner1, name) or
                        minetest.is_protected(corner2, name) or
                        minetest.is_protected(corner3, name) or
                        minetest.is_protected(corner4, name) or
                        minetest.is_protected(corner5, name) or
                        minetest.is_protected(corner6, name) or
                        minetest.is_protected(corner7, name) then
                return true
        end

        -- compute the optimal interval in every direction so that all borders are always checked.
        -- e.g. for a range 0..6 and requested interval 5, the actual interval: d[c], will be 2.9999, so that the coordinates 0, 3, 6 will be checked.
        -- e.g. for a range 0..9 and requested interval 5, the actual interval: d[c], will be 4.4999, so that the coordinates 0, 4, 9 will be checked.
        -- Substracting 1e-4 ensures that the max coordinate will always be reached by the for loop (which might otherwise not be the case due to rounding errors).
        -- This scheme (substracting 1e-4) fails for huge distances (approx. > 5000 * interval). It's safe to assume that that does not happen.
        d = {}
        for _,c in pairs({"x", "y", "z"}) do
            if maxp[c] >= minp[c] then
                d[c] = (maxp[c] - minp[c]) / math.ceil((maxp[c] - minp[c]) / interval) - 1e-4
            else
                d[c] = - (maxp[c] - minp[c]) / math.floor((maxp[c] - minp[c]) / interval) + 1e-4
            end
        end

        for zf = minp.z, maxp.z, d.z do
        for yf = minp.y, maxp.y, d.y do
        for xf = minp.x, maxp.x, d.x do
                local x = math.floor(xf + 0.5)
                local y = math.floor(yf + 0.5)
                local z = math.floor(zf + 0.5)
                local checkpos = {x = x, y = y, z = z}
                if checkpos ~= corner1 and
                                checkpos ~= corner2 and
                                checkpos ~= corner3 and
                                checkpos ~= corner4 and
                                checkpos ~= corner5 and
                                checkpos ~= corner6 and
                                checkpos ~= corner7 then
                        if minetest.is_protected(checkpos, name) then
                                return true
                        end
                end
        end
        end
        end

        return false
end
```

t4im commented:
Github tip: You can add syntax highlighting by appending the language to the code environment.
```lua
`​`​`lua
function default.intersects_protection(minp, maxp, name, interval)
end
`​`​`
```

paramat commented:
Rogier-5 interesting, with interval optimised like this we wouldn't have to check the corners first right? The loop would hit all 8 corners.

Rogier-5 commented:
> Rogier-5 interesting, with interval optimised like this we wouldn't have to check the corners first right?

>The loop would hit all 8 corners.

Indeed it would, however, it is much more efficient to check the corners first. Assuming protected areas are relatively large on average, if a tree will violate protection, then most likely, at least one of the corners will also violate protection. Checking that first saves all the checks of the interior area. That does not make much of a difference for regular trees, but it does for larger trees, e.g. from moretrees. A 20x30x20 tree (not the largest) would require up to 175 protection checks. If 167 of those can be saved by checking the corners first, that's a big improvement. Especially as protection checks can be expensive.

> Looks like the specified interval becomes the max interval used.

Indeed. The actual interval will be as large as possble, but not larger than the requested interval.

> I think we can assume maxp > minp.

That will most probably be true indeed. However, one never knows how somebody will use the function in the future. Allowing for maxp < minp is not very expensive, and it may save some headaches.

Rogier-5 commented:
I just spotted a bug in my code. The line:
```lua
if maxp[c] >= minp[c] then
```
should become:
```lua
if maxp[c] == minp[c] then
	d[c] = 1		-- Any value larger than 0 will do
elseif maxp[c] > minp[c] then
```

paramat commented:
Updated.

> Assuming protected areas are relatively large on average, if a tree will violate protection, then most likely, at least one of the corners will also violate protection. Checking that first saves all the checks of the interior area.

I decided to make it simple. Protected areas are often small and since trees are fairly tall it's quite likely the corners will be exterior to a protected area, especially when someone is trying to grief by planting a sapling underneath.
I decided to assume maxp > minp, it may not work well if that mistake is made but too bad. However we could add a warning, i would rather have a warning than allowing it to work with a mistake, leaving the mistake unnoticed.

Updated the code for appletree sapling with checks for buildable-to and a protection check at sapling position. Once it's satisfactory maybe i'll make another function to avoid unnecessary duplicating of code for every sapling.

DonBatman commented:
Just a thought. If trees had stages of groth then the checks wound be less intrusive. If each tree had 3 schematics, each a 3rd of the height of the full and timed so they grow in 3rds. This would allow for more checks without too many at a time.

Rogier-5 commented:
> Just a thought. If trees had stages of groth then the checks wound be less intrusive. If each tree had 3 schematics, each a 3rd of the height of the full and timed so they grow in 3rds. This would allow for more checks without too many at a time.

Maybe, maybe not. Several hundred protection checks when placing a single sapling is quite a lot, and they will compete with other server tasks, like water spreading, ABMs and timers. OTOH, as long as it's a player placing the sapling, and as long as there aren't dozens of players doing the same simultaneously, the impact on the server is probably limited, as a player usually can't place more than 1 sapling a second on average.

If the protection check happens at sapling growing time, then the impact can be much greater. Often, large numbers of simultaneous growing events happen as a block that has not been loaded for some time is loaded, and all accumulated timers expire simultaneously, and all accumulated ABMs fire simultaneously.

unlikely to be a real problem when placing a sapling, unless dozens of players are doing so at the same time. After all, a player can only place a limited number of saplings a minute.

Rogier-5 commented:
> I decided to make it simple. Protected areas are often small and since trees are fairly tall it's quite likely the corners will be exterior to a protected area, especially when someone is trying to grief by planting a sapling underneath. The code is much shorter and there are no checks for already-checked corners.

After some more thinking about it, I agree that it is indeed better to make the code shorter and simpler. Although I don't think protected areas will predominantly be smaller than even a larger tree: people will not only protect their build, but also some space around it, as well as sufficient air-space above it, this is really irrelevant, as the most common use-case is probably not where there is a protection violation (either by accident or by griefing), but where there isn't: a player placing a sapling in an unprotected area, or in an area he owns. And that will still cause the full area to be checked, only to conclude that no protection applies...

> I decided to assume maxp > minp, it may not work well if that mistake is made but too bad. However we could add a warning, i would rather have a warning than allowing it to work with a mistake, leaving the mistake unnoticed.

maxp, minp could be redefined as being two opposite corners of the area, instead of a minimum and maximum corner. Then either case would be valid. Whichever way, I personally don't really mind, but in any case, I always try to avoid things failing silently. So issuing a warning sounds like a good idea. Maybe also as a chat message to the user? You'd probably also have to return true, so that the sapling will not accidentally be placed in a protected area as a result of such a mistake. (it's better to get complaints about not being able to place saplings, than to get complaints about builds having been griefed).

Rogier-5 commented on mods/default/nodes.lua:
```diff
@@ -500,16 +500,58 @@ minetest.register_node("default:sapling", {
+		local player_name = placer:get_player_name()
+		-- Check sapling position for protection
+		if minetest.is_protected(pos, player_name) then
+			minetest.chat_send_player(player_name, "Node is protected")
```
Isn't there supposed to be a call to `core.record_protection_violation()` too ?

paramat commented:
Rogier-5 agreed.

Growing trees in stages: idea is fine in isolation but adds much complexity and a single stage is acceptable, so i oppose that for mtgame.

paramat commented:
Updated. 'on place' code is only updated for appletree sapling for now.
I can't easily test this so please can people test using appletree saplings?

paramat commented:
Updated and now complete.
Added a global 'sapling on place' function to avoid duplicated code in nodes.lua.
Testing is needed.

paramat commented:
Updated.
"Add a global 'sapling on place' function to avoid duplicated code in
nodes.lua."
Testing is needed.

paramat commented:
Now tested.

t4im commented on mods/default/trees.lua:
```diff
@@ -426,3 +427,48 @@ function default.grow_new_aspen_tree(pos)
+		sapling_name, minp_relative, maxp_relative, interval)
+	-- Position of sapling
+	local pos = pointed_thing.under
+	local node = minetest.get_node(pos)
```
First of all, that `on_place` function is pretty good!

The only improvement i can think of is handling `CONTENT_IGNORE`, which here is simply done by using:

```lua
local node = minetest.get_node_or_nil(pos)
local pdef = node and minetest.registered_nodes[node.name]
```

It's a rare condition, usually only happening on laggy servers, and even there only directly after teleportation, `/home` etc.
But when it would happen `get_node` would return a `buildable_to` node even without it being `buildable_to`, which would cause `pointed_thing.under` being replaced with a sapling later.
`get_node_or_nil` simply returns `nil` for `CONTENT_IGNORE`

paramat replied:
This may be true, but most other code in MT, even in builtin doesn't safegaurd against this.

paramat replied:
As you point out this is highly unlikely to happen, players do not tend to place saplings into ignore, and then it doesn't cause crash, so i think low priority.

SmallJoker replied:
Nothing is impossible on laggy servers. Adding that ignore check will make it totally fail-safe.

rubenwardy replied:
Doesn't get node return name = ignore? So it wouldn't crash, it would just exit below on that if not pdef

paramat replied:
But ignore does have a pdef.

SmallJoker commented on mods/default/trees.lua:
```diff
@@ -426,3 +427,48 @@ function default.grow_new_aspen_tree(pos)
+	local player_name = placer:get_player_name()
+	-- Check sapling position for protection
+	if minetest.is_protected(pos, player_name) then
+		minetest.chat_send_player(player_name, "Node is protected")
```
The protection mods already notify the player in `on_protection_violation`.

paramat replied:
Yeah i noticed, but teh extra information is useful to know whether the node is protected or the resulting tree will intersect protection.

paramat replied:
Since i provide my own message is 'minetest.record_protection_violation' needed at all?

paramat replied:
```
function core.record_protection_violation(pos, name)
	for _, func in pairs(core.registered_on_protection_violation) do
		func(pos, name)
	end
end
```
Might be best to leave it in there.

SmallJoker replied:
Right, the player might only have placed the node by accident. It's only a sapling so `minetest.record_protection_violation` could be removed.

SmallJoker commented on mods/default/trees.lua:
```diff
@@ -426,3 +427,48 @@ function default.grow_new_aspen_tree(pos)
+			itemstack:take_item()
+		end
+	else
+		minetest.chat_send_player(player_name, "Tree will intersect protection")
```
For same reason as above, this line should be moved below `record_protection_violation`. So the reason why the protection is violated follows the violation message, which looks better IMO.

paramat replied:
Agreed,

sofar commented:
I'm not a fan. Instead, saplings shouldn't grow or they should grow nodes in unprotected areas. Prohibiting placement of saplings in adjacent nodes that people own prevents them from growing decorative plants or simply growing some trees near their own protected areas' edges.

I feel this is trying to hard.

paramat commented:
Checks at sapling grow time would need to store placer data as metadata, this method here is simpler and consistent with other protection behaviour because it prevents placement if there is (or will be) a violation.

> they should grow nodes in unprotected areas

There would be half trees and trees without trunks etc., better to have no tree at all.
That would also need per-node protection checks which is far too slow.

> Prohibiting placement of saplings in adjacent nodes that people own prevents them from growing decorative plants or simply growing some trees near their own protected areas' edges.

If such a tree would violate protection that shouldn't be allowed. Decorative saplings are not that important. Both would be placeable by moving a mere 2 nodes away from another player's protected area.

This is less complex than the initial suggestion of checking every node. It tries hard because this is actually a serious griefing issue, see Shara's comments a few days back in project channel about a major griefing incident.

paramat commented:
Please read the IRC dev channel discussions about this, they will explain how we decided on what to do. At first i didn't think it very important and was suggesting a less robust solution. The thoroughness is configurable using 'interval', currently most trees will have 12 or 16 checks, acacia 27.

paramat commented on mods/default/functions.lua:
```diff
@@ -481,3 +481,43 @@ minetest.register_abm({
+	for zf = minp.z, maxp.z, d.z do
+	for yf = minp.y, maxp.y, d.y do
+	for xf = minp.x, maxp.x, d.x do
+		local x = math.floor(xf + 0.5)
```
So it seems i can move these roundings outward from the inner loops, will do tthis.

paramat commented:
I'll move the roundings outward from the inner loops.

Rogier-5 commented:
> I'm not a fan. Instead, saplings shouldn't grow or they should grow nodes in unprotected areas. 

Apart from the fact that saplings would need to have a kind of 'owner', the feature of allowing them to grow only in 'allowed' areas requires the tree generating code to check the protection of every node at the moment a tree grows, and to only place nodes where possible. For a moretrees tree for example, that could easily be 1000 or more protection checks. Even up to several 1000 for the largest trees. As protection checks are not exceptionally fast, and probably at least one or two orders of magnitude slower than just placing the node without protection check, there would be a significant performance impact. In addition, as ABMs and timers fire more often at block load time, a significant number of saplings may grow at approximately the same time. Multiplied by 1000 or more protection checks each, there would probably be significant lag.

> Prohibiting placement of saplings in adjacent nodes that people own prevents them from growing decorative plants or simply growing some trees near their own protected areas' edges.

Only if adjacent to another player's protected area, and if the grown tree would grief the other player's protected area.

Rogier-5 commented:
> Decorative saplings are not that important.

And if they *are* considered important, I'm sure somebody won't hesitate to create or adapt a mod to craft decorative saplings from real saplings.

paramat commented:
Updated.

Moved these rounding lines outward from the innerloops to avoid unnecessary calculations.
```
+	for zf = minp.z, maxp.z, d.z do
+		local z = math.floor(zf + 0.5)
+		for yf = minp.y, maxp.y, d.y do
+			local y = math.floor(yf + 0.5)
+			for xf = minp.x, maxp.x, d.x do
+				local x = math.floor(xf + 0.5)
+				if minetest.is_protected({x = x, y = y, z = z}, player_name) then
+					return true
+				end
+			end
+		end
+	end
```
Because the violation of the node sapling is placed on is a standard violation already notified through 'minetest.record_protection_violation(pos, player_name)' i removed the unnecessary second notification.
However the second message is kept for the tree violation to explain why placement was refused, the second message is moved to actually be the second message.

paramat commented:
Tested.

paramat commented:
Updated first post with explanation and links to discussion @sofar

t4im commented:
Tested this, too and it works.

Protection checks are done, where the stone is:

![treeprotection](https://cloud.githubusercontent.com/assets/1937987/17190601/be80e77e-5446-11e6-9255-b37dce0ed9cc.png)

paramat commented:
Good as expected.
Since this is urgently needed, has been discussed, developed and reviewed by many people, and has been tested by 2 people, i will now merge this. Sofar is not keen and other devs are absent but i think it's reasonable for this to go ahead on my approval (it also has approval from a possible future dev).

paramat commented:
0ac096991c51d664007ebb53b8117baca0271459
